File: kluctl/cli/command_result.py
import dataclasses
import os

import click
from deepdiff.helper import NotPresent

from kluctl.cli.utils import build_seen_images
from kluctl.deployment.deployment_collection import CommandResult
from kluctl.diff.k8s_diff import unified_diff_object, changes_to_yaml
from kluctl.utils.exceptions import CommandError
from kluctl.utils.k8s_object_utils import get_long_object_name, get_long_object_name_from_ref, get_object_ref
from kluctl.utils.pretty_table import pretty_table
from kluctl.utils.yaml_utils import yaml_dump, yaml_dump_all


def format_command_result_text(command_result: CommandResult):
    result = ''

    if command_result.warnings:
        result += "\nWarnings:\n"
        result += pretty_errors(command_result.warnings)

    if command_result.new_objects:
        result += "\nNew objects:\n"
        for x in command_result.new_objects:
            result += "  %s\n" % get_long_object_name(x, include_api_version=False)

    if command_result.changed_objects:
        result += "\nChanged objects:\n"
        for x in command_result.changed_objects:
            result += "  %s\n" % get_long_object_name(x["new_object"], include_api_version=False)

        result += "\n"
        for x in command_result.changed_objects:
            object = x["new_object"]
            changes = x["changes"]
            result += "%s" % pretty_changes(get_object_ref(object), changes)

    if command_result.deleted_objects:
        result += "\nDeleted objects:\n"
        for x in command_result.deleted_objects:
            result += "  %s\n" % get_long_object_name_from_ref(x)

    if command_result.hook_objects:
        result += "\nApplied hooks:\n"
        for x in command_result.hook_objects:
            result += "  %s\n" % get_long_object_name(x)

    if command_result.orphan_objects:
        result += "\nOrphan objects:\n"
        for ref in command_result.orphan_objects:
            result += "  %s\n" % get_long_object_name_from_ref(ref)

    if command_result.errors:
        result += "\nErrors:\n"
        result += pretty_errors(command_result.errors)

    return result

def pretty_errors(errors):
    table = [("Object", "Message")]
    for e in errors:
        table.append((get_long_object_name_from_ref(e.ref), e.message))
    return pretty_table(table, [60])

def pretty_changes(ref, changes):
    ret = 'Diff for object %s\n' % get_long_object_name_from_ref(ref)

    table = [('Path', 'Diff')]
    for c in changes:
        if "unified_diff" in c:
            diff = c["unified_diff"]
        else:
            diff = unified_diff_object(c.get("old_value", NotPresent()), c.get("new_value", NotPresent()))
        table.append((c["path"], diff))

    ret += pretty_table(table, [60])

    return ret

def format_command_result_yaml(c, command_result: CommandResult):
    result = {
        "diff": changes_to_yaml(command_result.new_objects, command_result.changed_objects),
        "deleted_objects": command_result.deleted_objects,
        "applied_hooks": command_result.hook_objects,
        "orphan_objects": [{"ref": dataclasses.asdict(ref)} for ref in command_result.orphan_objects],
        "errors": [dataclasses.asdict(x) for x in command_result.errors],
        "warnings": [dataclasses.asdict(x) for x in command_result.warnings],
        "images": build_seen_images(c, True),
    }
    return yaml_dump(result)

def format_command_result(c, command_result, format):
    if format == "text":
        return format_command_result_text(command_result)
    elif format == "yaml":
        return format_command_result_yaml(c, command_result)
    else:
        raise CommandError(f"Invalid format: {format}")

def format_validate_result(result, format):
    if format == "text":
        str = ""
        if result.warnings:
            str += "Validation Warnings:\n"
            for item in result.warnings:
                str += "  %s: reason=%s, message=%s\n" % (get_long_object_name_from_ref(item.ref), item.reason, item.message)
        if result.errors:
            if str:
                str += "\n"
            str += "Validation Errors:\n"
            for item in result.errors:
                str += "  %s: reason=%s, message=%s\n" % (get_long_object_name_from_ref(item.ref), item.reason, item.message)
        if result.results:
            if str:
                str += "\n"
            str += "Results:\n"
            for item in result.results:
                str += "  %s: reason=%s, message=%s\n" % (get_long_object_name_from_ref(item.ref), item.reason, item.message)
        return str
    if format == "yaml":
        y = yaml_dump(dataclasses.asdict(result))
        return y
    else:
        raise CommandError(f"Invalid format: {format}")


def output_command_result(output, c, command_result):
    if not output:
        output = ["text"]
    for o in output:
        s = o.split("=", 1)
        format = s[0]
        path = None
        if len(s) > 1:
            path = s[1]
        s = format_command_result(c, command_result, format)
        output_result(path, s)


def output_validate_result(output, result):
    if not output:
        output = ["text"]
    for o in output:
        s = o.split("=", 1)
        format = s[0]
        path = None
        if len(s) > 1:
            path = s[1]
        s = format_validate_result(result, format)
        output_result(path, s)


def output_yaml_result(output, result, all=False):
    output = output or [None]
    if all:
        s = yaml_dump_all(result)
    else:
        s = yaml_dump(result)
    for o in output:
        output_result(o, s)


def output_result(output_file, result):
    path = None
    if output_file and output_file != "-":
        path = os.path.expanduser(output_file)
    if path is None:
        click.echo(result)
    else:
        with open(path, "wt") as f:
            f.write(result)

File: cmd/kluctl/commands/command_result.go
package commands

import (
	"bytes"
	"context"
	"fmt"
	"github.com/kluctl/kluctl/v2/cmd/kluctl/args"
	"github.com/kluctl/kluctl/v2/pkg/diff"
	"github.com/kluctl/kluctl/v2/pkg/status"
	"github.com/kluctl/kluctl/v2/pkg/types/k8s"
	"github.com/kluctl/kluctl/v2/pkg/types/result"
	"github.com/kluctl/kluctl/v2/pkg/utils"
	"github.com/kluctl/kluctl/v2/pkg/yaml"
	"io"
	"os"
	"strings"
)

func formatCommandResultText(cr *result.CommandResult, short bool) string {
	buf := bytes.NewBuffer(nil)

	if len(cr.Warnings) != 0 {
		buf.WriteString("\nWarnings:\n")
		prettyErrors(buf, cr.Warnings)
	}

	var newObjects []k8s.ObjectRef
	var changedObjects []k8s.ObjectRef
	var deletedObjects []k8s.ObjectRef
	var orphanObjects []k8s.ObjectRef
	var appliedHookObjects []k8s.ObjectRef

	for _, o := range cr.Objects {
		if o.New {
			newObjects = append(newObjects, o.Ref)
		}
		if len(o.Changes) != 0 {
			changedObjects = append(changedObjects, o.Ref)
		}
		if o.Deleted {
			deletedObjects = append(deletedObjects, o.Ref)
		}
		if o.Orphan {
			orphanObjects = append(orphanObjects, o.Ref)
		}
		if o.Hook {
			appliedHookObjects = append(appliedHookObjects, o.Ref)
		}
	}

	if len(newObjects) != 0 {
		buf.WriteString("\nNew objects:\n")
		prettyObjectRefs(buf, newObjects)
	}
	if len(changedObjects) != 0 {
		buf.WriteString("\nChanged objects:\n")
		prettyObjectRefs(buf, changedObjects)

		if !short {
			buf.WriteString("\n")
			for i, o := range cr.Objects {
				if len(o.Changes) == 0 {
					continue
				}
				if i != 0 {
					buf.WriteString("\n")
				}
				prettyChanges(buf, o.Ref, o.Changes)
			}
		}
	}

	if len(deletedObjects) != 0 {
		buf.WriteString("\nDeleted objects:\n")
		prettyObjectRefs(buf, deletedObjects)
	}

	if len(appliedHookObjects) != 0 {
		buf.WriteString("\nApplied hooks:\n")
		prettyObjectRefs(buf, appliedHookObjects)
	}
	if len(orphanObjects) != 0 {
		buf.WriteString("\nOrphan objects:\n")
		prettyObjectRefs(buf, orphanObjects)
	}

	if len(cr.Errors) != 0 {
		buf.WriteString("\nErrors:\n")
		prettyErrors(buf, cr.Errors)
	}

	return buf.String()
}

func prettyObjectRefs(buf io.StringWriter, refs []k8s.ObjectRef) {
	for _, ref := range refs {
		_, _ = buf.WriteString(fmt.Sprintf("  %s\n", ref.String()))
	}
}

func prettyErrors(buf io.StringWriter, errors []result.DeploymentError) {
	for _, e := range errors {
		prefix := ""
		if s := e.Ref.String(); s != "" {
			prefix = fmt.Sprintf("%s: ", s)
		}
		_, _ = buf.WriteString(fmt.Sprintf("  %s%s\n", prefix, e.Message))
	}
}

func prettyChanges(buf io.StringWriter, ref k8s.ObjectRef, changes []result.Change) {
	_, _ = buf.WriteString(fmt.Sprintf("Diff for object %s\n", ref.String()))

	var t utils.PrettyTable
	t.AddRow("Path", "Diff")

	for _, c := range changes {
		t.AddRow(c.JsonPath, c.UnifiedDiff)
	}
	s := t.Render([]int{60})
	_, _ = buf.WriteString(s)
}

func formatCommandResultYaml(cr *result.CommandResult) (string, error) {
	b, err := yaml.WriteYamlString(cr.ToCompacted())
	if err != nil {
		return "", err
	}
	return b, nil
}

func formatCommandResult(cr *result.CommandResult, format string, short bool) (string, error) {
	switch format {
	case "text":
		return formatCommandResultText(cr, short), nil
	case "yaml":
		return formatCommandResultYaml(cr)
	default:
		return "", fmt.Errorf("invalid format: %s", format)
	}
}

func prettyValidationResults(buf io.StringWriter, results []result.ValidateResultEntry) {
	var t utils.PrettyTable
	t.AddRow("Object", "Message")

	for _, e := range results {
		t.AddRow(e.Ref.String(), e.Message)
	}
	s := t.Render([]int{60})
	_, _ = buf.WriteString(s)
}

func formatValidateResultText(vr *result.ValidateResult) string {
	buf := bytes.NewBuffer(nil)

	if len(vr.Warnings) != 0 {
		buf.WriteString("\nValidation Warnings:\n")
		prettyErrors(buf, vr.Warnings)
	}

	if len(vr.Errors) != 0 {
		if buf.Len() != 0 {
			buf.WriteString("\n")
		}
		buf.WriteString("Validation Errors:\n")
		prettyErrors(buf, vr.Errors)
	}

	if len(vr.Results) != 0 {
		if buf.Len() != 0 {
			buf.WriteString("\n")
		}
		buf.WriteString("Results:\n")
		prettyValidationResults(buf, vr.Results)
	}

	if len(vr.Drift) != 0 {
		if buf.Len() != 0 {
			buf.WriteString("\n")
		}
		buf.WriteString("Drift:\n")
		for i, o := range vr.Drift {
			if len(o.Changes) == 0 {
				continue
			}
			if i != 0 {
				buf.WriteString("\n")
			}
			prettyChanges(buf, o.Ref, o.Changes)
		}
	}

	return buf.String()
}

func formatValidateResultYaml(vr *result.ValidateResult) (string, error) {
	b, err := yaml.WriteYamlString(vr)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func formatValidateResult(vr *result.ValidateResult, format string) (string, error) {
	switch format {
	case "text":
		return formatValidateResultText(vr), nil
	case "yaml":
		return formatValidateResultYaml(vr)
	default:
		return "", fmt.Errorf("invalid validation result format: %s", format)
	}
}

func outputHelper(ctx context.Context, output []string, cb func(format string) (string, error)) error {
	if len(output) == 0 {
		output = []string{"text"}
	}
	for _, o := range output {
		s := strings.SplitN(o, "=", 2)
		format := s[0]
		var path *string
		if len(s) > 1 {
			path = &s[1]
		}
		r, err := cb(format)
		if err != nil {
			return err
		}

		err = outputResult(ctx, path, r)
		if err != nil {
			return err
		}
	}
	return nil
}

func outputCommandResult(ctx *commandCtx, flags args.OutputFormatFlags, cr *result.CommandResult, writeToResultStore bool) error {
	status.Flush(ctx.ctx)

	if !flags.NoObfuscate {
		var obfuscator diff.Obfuscator
		err := obfuscator.ObfuscateResult(cr)
		if err != nil {
			return err
		}
	}

	var resultStoreErr error
	if writeToResultStore && ctx.resultStore != nil {
		s := status.Start(ctx.ctx, "Writing command result")
		defer s.Failed()
		resultStoreErr = ctx.resultStore.WriteCommandResult(cr)
		if resultStoreErr != nil {
			s.FailedWithMessage("Failed to write result to result store: %s", resultStoreErr.Error())
		} else {
			s.Success()
		}
	}

	err := outputHelper(ctx.ctx, flags.OutputFormat, func(format string) (string, error) {
		return formatCommandResult(cr, format, flags.ShortOutput)
	})
	if err == nil && resultStoreErr != nil {
		return resultStoreErr
	}
	return err
}

func outputValidateResult(ctx context.Context, output []string, vr *result.ValidateResult) error {
	status.Flush(ctx)

	return outputHelper(ctx, output, func(format string) (string, error) {
		return formatValidateResult(vr, format)
	})
}

func outputYamlResult(ctx context.Context, output []string, result interface{}, multiDoc bool) error {
	status.Flush(ctx)

	if len(output) == 0 {
		output = []string{"-"}
	}
	var s string
	if multiDoc {
		l, ok := result.([]interface{})
		if !ok {
			return fmt.Errorf("object is not a list")
		}
		x, err := yaml.WriteYamlAllString(l)
		if err != nil {
			return err
		}
		s = x
	} else {
		x, err := yaml.WriteYamlString(result)
		if err != nil {
			return err
		}
		s = x
	}
	for _, path := range output {
		err := outputResult(ctx, &path, s)
		if err != nil {
			return err
		}
	}
	return nil
}

func outputResult(ctx context.Context, f *string, result string) error {
	var w io.Writer
	w = getStdout(ctx)
	if f != nil && *f != "-" {
		f, err := os.Create(*f)
		if err != nil {
			return err
		}
		defer f.Close()
		w = f
	}
	_, err := w.Write([]byte(result))
	return err
}
